components/feature/app-links/src/main/java/mozilla/components/feature/app/links/AppLinksUseCases.kt

```diff
@@ -0,0 +1,147 @@
+            .toHashSet()
+    }
+
+    inner class GetAppLinkConfig internal constructor() {
+        fun invoke(url: String): AppLinkRedirect {
+            val intents = createBrowsableIntents(url)
+            val appIntent = intents.firstOrNull {
+                getNonBrowserActivities(it).isNotEmpty()
+            }
+
+            val webUrls = intents.mapNotNull {
+                if (it.data?.isHttpOrHttps == true) it.dataString else null
+            }
+
+            val webUrl = webUrls.firstOrNull { it != url } ?: webUrls.firstOrNull()
+
+            return AppLinkRedirect(appIntent, webUrl, webUrl != url)
+        }
+
+        private fun getNonBrowserActivities(intent: Intent): List<ResolveInfo> {
+            return findActivities(intent)
+                .filter { !browserPackageNames.contains(it.activityInfo.packageName) }
+        }
+
+        private fun createBrowsableIntents(url: String): List<Intent> {
+            val intent = Intent.parseUri(url, 0)
+
+            if (intent.action == Intent.ACTION_VIEW) {
+                intent.addCategory(Intent.CATEGORY_BROWSABLE)
+            }
+
+            return when (intent.data?.isHttpOrHttps) {
+                null -> emptyList()
+                true -> listOf(intent)
+                false -> {
+                    val fallback = intent.getStringExtra(EXTRA_BROWSER_FALLBACK_URL)?.let {
+                        createBrowsableIntents(it)
+                    } ?: emptyList()
+
+                    val marketplaceIntent = intent.`package`?.let {
+                        Intent.parseUri(MARKET_INTENT_URI_PACKAGE_PREFIX + it, 0)
+                    }
+
+                    return listOf(intent) + fallback + listOfNotNull(marketplaceIntent)
+                }
+            }
+        }
+    }
+
+    class OpenAppLinkUseCase internal constructor(
+        private val context: Context,
+        private val fragmentManager: FragmentManager,
+        private val dialog: RedirectDialogFragment
+    ) {
+        fun invoke(redirect: AppLinkRedirect, session: EngineSession): Boolean {
+            val intent = redirect.appIntent ?: return false
+
+            dialog.setAppLinkRedirect(redirect)
+            dialog.onConfirmRedirect = {
+                val openInIntent = Intent.createChooser(
+                    intent,
+                    context.getString(R.string.mozac_feature_applinks_open_in)
+                ).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
+
+                context.startActivity(openInIntent)
+            }
+
+            dialog.onDismiss(object : DialogInterface {
+                override fun dismiss() {
+                    redirect.webUrl?.let {
+                        session.loadUrl(it)
+                    }
+                }
+
+                override fun cancel() {
+                    dismiss()
+                }
+            })
+
+            dialog.show(fragmentManager, FRAGMENT_TAG)
+
+            // TODO check that the user has made a choice.
```

Comment:
I think we also have to take into consideration when the app gets kill by the OS, to reattach the callbacks. You can see an example on [DownloadsFeature](https://github.com/mozilla-mobile/android-components/blob/79ca1f4915b806d6292f57e527d93062ca036064/components/feature/downloads/src/main/java/mozilla/components/feature/downloads/DownloadsFeature.kt#L59).

Reply:
:+1: